service/debugger/debugger.go

```diff
@@ -0,0 +1,519 @@
+	var createdBp *api.BreakPoint
+	err := d.withProcess(func(p *proctl.DebuggedProcess) error {
+		var loc string
+		switch {
+		case len(requestedBp.File) > 0:
+			loc = fmt.Sprintf("%s:%d", requestedBp.File, requestedBp.Line)
+		case len(requestedBp.FunctionName) > 0:
+			loc = requestedBp.FunctionName
+		default:
+			return fmt.Errorf("no file or function name specified")
+		}
+
+		bp, breakError := p.BreakByLocation(loc)
+		if breakError != nil {
+			return breakError
+		}
+		createdBp = convertBreakPoint(bp)
+		return nil
+	})
+	return createdBp, err
+}
+
+func (d *Debugger) DeleteBreakPoint(requestedBp *api.BreakPoint) (*api.BreakPoint, error) {
+	var deletedBp *api.BreakPoint
+	err := d.withProcess(func(p *proctl.DebuggedProcess) error {
+		bp, err := p.Clear(requestedBp.Addr)
+		if err != nil {
+			return fmt.Errorf("Can't clear breakpoint @%x: %s", requestedBp.Addr, err)
+		}
+		deletedBp = convertBreakPoint(bp)
+		return nil
+	})
+	return deletedBp, err
+}
+
+func (d *Debugger) BreakPoints() []*api.BreakPoint {
+	bps := []*api.BreakPoint{}
+	d.withProcess(func(p *proctl.DebuggedProcess) error {
+		for _, bp := range p.HWBreakPoints {
+			if bp == nil {
+				continue
+			}
+			bps = append(bps, convertBreakPoint(bp))
+		}
+
+		for _, bp := range p.BreakPoints {
+			if bp.Temp {
+				continue
+			}
+			bps = append(bps, convertBreakPoint(bp))
+		}
+		return nil
+	})
+	return bps
+}
+
+func (d *Debugger) FindBreakPoint(id int) *api.BreakPoint {
+	for _, bp := range d.BreakPoints() {
+		if bp.ID == id {
+			return bp
+		}
+	}
+	return nil
+}
+
+func (d *Debugger) Threads() []*api.Thread {
+	threads := []*api.Thread{}
+	d.withProcess(func(p *proctl.DebuggedProcess) error {
+		for _, th := range p.Threads {
+			threads = append(threads, convertThread(th))
+		}
+		return nil
+	})
+	return threads
+}
+
+func (d *Debugger) FindThread(id int) *api.Thread {
+	for _, thread := range d.Threads() {
+		if thread.ID == id {
+			return thread
+		}
+	}
+	return nil
+}
+
+// Command handles commands which control the debugger lifecycle. Like other
+// debugger operations, these are executed one at a time as part of the
+// process operation pipeline.
+//
+// The one exception is the Halt command, which can be executed concurrently
+// with any operation.
+func (d *Debugger) Command(command *api.DebuggerCommand) (*api.DebuggerState, error) {
+	var err error
+	switch command.Name {
+	case api.Continue:
+		err = d.withProcess(func(p *proctl.DebuggedProcess) error {
+			e := p.Continue()
+			return e
+		})
+	case api.Next:
+		err = d.withProcess(func(p *proctl.DebuggedProcess) error {
+			return p.Next()
+		})
+	case api.Step:
+		err = d.withProcess(func(p *proctl.DebuggedProcess) error {
+			return p.Step()
+		})
+	case api.SwitchThread:
+		err = d.withProcess(func(p *proctl.DebuggedProcess) error {
+			return p.SwitchThread(command.ThreadID)
+		})
+	case api.Halt:
+		// RequestManualStop just sends a signal, so it's safe to access the process directly.
```

Comment:
Change to:

`// RequestManualStop does not invoke any ptrace syscalls, so it's safe to access the process directly.`